--- /usr/ports/Mk/bsd.port.mk.prepkgng	2011-01-14 00:14:59.962893785 +0100
+++ /usr/ports/Mk/bsd.port.mk	2011-01-14 18:12:25.129947229 +0100
@@ -2493,6 +2493,12 @@
 
 .if !defined(PKG_ARGS)
 PKG_ARGS=		-v -c -${COMMENT:Q} -d ${DESCR} -f ${TMPPLIST} -p ${PREFIX} -P "`cd ${.CURDIR} && ${MAKE} actual-package-depends | ${GREP} -v -E ${PKG_IGNORE_DEPENDS} | ${SORT} -u -t : -k 2`" ${EXTRA_PKG_ARGS} $${_LATE_PKG_ARGS}
+.if defined(PKGNG)
+PKG_ARGS+= -n ${PKGNAME}
+.if exists(${PKGMESSAGE})
+PKG_ARGS+= -M ${PKGMESSAGE}
+.endif
+.endif
 .if !defined(NO_MTREE)
 PKG_ARGS+=		-m ${MTREE_FILE}
 .endif
@@ -5552,6 +5558,7 @@
 		fi; \
 	done
 
+.if !defined(PKGNG)
 ACTUAL-PACKAGE-DEPENDS?= \
 	if [ "${_LIB_RUN_DEPENDS}" != "  " ]; then \
 		origins=$$(for pkgname in ${PKG_DBDIR}/*; do \
@@ -5579,8 +5586,21 @@
 				shift 2; \
 			done; \
 		done; \
-		[ -z "$$packagelist" ] || ${AWK} -F '( |:)' 'BEGIN { pkgname="broken_contents" } /@pkgdep / { pkgname=$$2 } /@comment DEPORIGIN:/ { printf "%s:%s\n", pkgname, $$3; pkgname="broken_contents" }' $$packagelist; \
+		[ -z "$$packagelist" ] || ${AWK} -F '( |:)' 'BEGIN { pkgname="broken_contents" } /@pkgdep / { pkgname=$$3 } /@comment DEPORIGIN:/ { printf "%s:%s\n", pkgname, $$3; pkgname="broken_contents" }' $$packagelist; \
 	fi
+.else
+ACTUAL-PACKAGE-DEPENDS?= \
+	if [ "${_LIB_RUN_DEPENDS}" != "  " ]; then \
+		for dir in ${_LIB_RUN_DEPENDS:C,[^:]*:([^:]*):?.*,\1,}; do \
+			pkgname=$$(${PKG_INFO} -q $${dir\#\#${PORTSDIR}/}); \
+			${ECHO_CMD} $$pkgname:$${dir\#\#${PORTSDIR}/}; \
+			for pkg in $$(${PKG_INFO} -d $${dir\#\#${PORTSDIR}/}); do\
+				origin=$$(${PKG_INFO} -qo $${pkg%-*}); \
+				${ECHO_CMD} $$pkg:$$origin; \
+			done; \
+		done; \
+	fi
+.endif
 
 # Print out package names.
 
@@ -5993,6 +6013,7 @@
 	fi
 	@${RM} -rf ${PKG_DBDIR}/${PKGNAME}
 .endif
+.if !defined(PKGNG)
 	@if [ ! -d ${PKG_DBDIR}/${PKGNAME} ]; then \
 		${ECHO_MSG} "===>   Registering installation for ${PKGNAME}"; \
 		${MKDIR} ${PKG_DBDIR}/${PKGNAME}; \
@@ -6031,6 +6052,11 @@
 		${RM} -f /tmp/${PKGNAME}-required-by; \
 	fi
 .else
+	@${ECHO_MSG} "===>   Registering installation for ${PKGNAME}"
+
+	@${PKG_CMD} ${PKG_ARGS}
+.endif
+.else
 	@${DO_NADA}
 .endif
 .endif
